### DISCUSSION TITLE
What are `Workspace`'s supposed to do?

### DISCUSSION THREAD

**Discussion by m10d** [flxzt/rnote#833] [Q&A]
I'd like to submit a bug/feature-request, but to make it meaningful I want to understand what workspaces are.  I will cut/paste/summarize any responses given here so that any effort given in responding here is not lost.  Thank you for RNote, I'm optimistic this will let me kick the microsoft surface windows install, finally (only holdout is OneNote).

So:

2.  I install via Flatpak, as recommended.  [I totally support this, btw - more encapsulation of apps is absolutely the way we should be going.]
1. grant the installed RNote flatpak file permissions to a folder `/home/myuser/my-sync-folder/rnote-school`  (happens to be a `syncthing` share folder I created for it, and a bit more convoluted path as I have a lot of shares between my family and organization, broken up quite a bit so that disk usage can be minimized on mobiles/tablets)
3. create an `RNote workspace` in that folder
4. ... I go about my business, not really thinking about `rnote` *_config_* and what not
5. some time later I create a new folder _from rnote_, along with a couple new notebooks inside that folder.  Add some important content, save, leave
6. go to present that information from my laptop (synced via syncthing)

but **OOPS** I accidentally navigated out of the workspace, saving the file _-somewhere-_ else:

- by creating a `workspace`, as a user I expected all folders & files I created to be _within_ that workspace folder.  
    - FWIW I fully agree with not "abstracting" the filesystem - but I expected `workspace` to define a top-level ~handle to that [filesystem] `tree`
- since `RNote` happily let me violate that, the notes are _not_ "part" of a the workspace as I thought they would, and were thus not synced.    
    - This has some similarities to https://github.com/flxzt/rnote/issues/777 
- Additionally_ thanks to `flatpak` encapsulation, the folders/files I created are _**not** even **in** my normal home dir_.  I'm sure I can find them - but a quick `find ~ ...` didn't turn up anything. 

Poking around, as best I can tell, `workspace` today is just a file browser, which when opened remembers the last open directory?

**Comment by flxzt** [comment]
Yes, the workspaces are essentially folder bookmarks with some customization (color, name, icon). But they also integrate with the save/load dialogs. They should always pick the folder of the selected workspace by default.

> but I expected workspace to define a top-level ~handle to that [filesystem] tree

I am not sure I understand what you mean here.

> the folders/files I created are not even in my normal home dir

It depends what you as the user has selected, doesn't it? From a user experience POV ideally if a directory outside of the flatpak granted ones is picked, a "grant permissiosns" dialog should be displayed, but I don't think this feature exists for flatpak portals.

**Reply by m10d** [reply]
what I don't understand, is why is the user giving the workspace a per-workspace location, if creating&saving files&directories is not at all linked to that location?  That's what I mean by top-level handle; the way the UX works (and the way I hope it would function) is that the workspace is a tree, with the `Directory` as given on the `Edit Workspace` modal is the `root` of the tree - traversing "up" past the root is (should be) impossible

Since workspaces _are_ a view of the filesystem (I support this - adding an abstraction would increase complexity w/out adding value), I feel doing anything **_except_** this (limit to _subdirectories_ of `root`) is not meaningful since the "workspace" in that case would add nothing except for a shortcut to file-browse "last open filesystem directory".  Is it adding anything else I'm missing?

While from a user perspective I feel the above is least-surprise and valuable use case, this is also supported by the reality of the recommended install method.  As you say, I do believe Flatpak encapsulation is "transparent" to the user, so there is no way (??) the app could know & warn that the user is making a file they may never be able to find outside of the running app - within this reality rnote is allowing a really bad result.

**Comment by m10d** [comment]
> They should always pick the folder of the selected workspace by default.

Ok, I do see this.  Defaulting to the "workspace directory" does seem correct/optimal.  I see that rnote uses the currently selected workspace, not the one that was selected when the document was created.  Today there is no automatic-save of new documents/notebooks (that would be a new feature AFAICT), this seems logical.  If that feature were implemented, this defaulting to workspace directory could be moot, or it could prompt the user "notebook saved in default location `/the/workspaceXX/folder/z.rnote`, would you like to move into `/some/workspaceYY` ".  Obviously a completely different feature, just wanted to acknowledge it.

Doesn't the fact that rnote defaults to save at workspace root just reinforce my perspective on helping the user follow their intent?  If the user wants a workspace containing all possible save locations, they can define the workspace folder at `/`.  rnote already appears to store `workspace identity-file`'s (or however the workspace state is stored) somewhere central (eg not in the `workspace folder`), so this would seem to have zero usability tradeoffs, and a significant upside?

**Reply by flxzt** [reply]
A possible recovery/autosave drafts feature should work outside of the current selected workspace, thats true. And still prompt the user when they save explicitely, therefore making a conscious choice by selecting some directory.

I still am unsure if I understand you correctly, but in essence: are you suggesting that the current workspace should act as the root directory for all save/load dialogs, so essentially overlaying a filesystem or abstracting it away in some way? I mean, I can see the argument for it, but I believe that would confuse like 99% of all users.

Just to give some perspective on how I intended it to be use: my personal workflow is: I have a workspace for my University courses (called "Uni"), and depending on the current course I am visiting, I'll quickly navigate within Rnote to this subdirectory (thats why it is possible to traverse through opening folders and that's also why there is a "move dir up" button). So I'll constantly change the workspace dir depending on the course. With this method, I have only one workspace for all of my university stuff, I can access rnote files quickly because all of them are in reach of one or two folders and new notes are all saved in the correct folder most of the time.

I appreciate the discussion though, getting behaviour like this correct is what improves the app a lot :)

**Comment by Fra0815** [comment]
Workspace logic: if I create a new file, with auto save enable I expect the file to be saved in the directory of the Workspace, If I want to personalize the file location instead of clicking once on the plus button, by clicking that button twice I should be able to select a different directory.

**Reply by Fra0815** [reply]
Instead the default directory is the last used regardless the workspace.

**Reply by Fra0815** [reply]
Or better yet, just let the user to define what is workspace dependent and not, or to create group of workspaces and defining their behaviour. For example, I want to the saving directory te be work space dependent, while I would like the layout to be workspace independent. Maybe someone else prefer to personalize the layout for each workspace

**Reply by flxzt** [reply]
when a file is saved on the first time, the directory is always used for subsequent auto saves. I don't think it's desirable to change the save directory once it is established.
